python/ray/experimental/channel/common.py

```diff
@@ -333,23 +352,57 @@ def start(self):
         pass
 
     def _read_list(self, timeout: Optional[float] = None) -> List[Any]:
-        results = []
-        for c in self._input_channels:
-            start_time = time.monotonic()
-            results.append(c.read(timeout))
-            if timeout is not None:
-                timeout -= time.monotonic() - start_time
-                timeout = max(timeout, 0)
+        # Release the buffers of the channels that were not read in the last
+        # `read` call because a RayTaskError was found. If we don't do this,
+        # the read operation will read different versions of the object refs.
+        if self._unread_channels is not None:
+            self._release_channel_buffers(self._unread_channels, timeout)
+            # Do not update `remaining_timeout` to avoid causing an unexpected
+            # timeout in the following read operation.
+            self._unread_channels = None
+
+        results = [None for _ in range(len(self._input_channels))]
+        if timeout is None or timeout == -1:
```

Comment:
The default timeout (`None`) is infinite which is the same when `timeout` is -1.